src/sql/src/query_model/mod.rs

```diff
@@ -0,0 +1,417 @@
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::cell::RefCell;
+use std::collections::BTreeSet;
+use std::collections::{HashMap, HashSet};
+use std::fmt;
+
+mod dot_generator;
+mod hir_generator;
+mod scalar_expr;
+#[cfg(test)]
+mod test;
+
+pub use scalar_expr::*;
+
+pub type QuantifierId = usize;
+pub type BoxId = usize;
+pub type QuantifierSet = BTreeSet<QuantifierId>;
+
+/// A Query Graph Model instance represents a SQL query.
+#[derive(Debug)]
+pub struct Model {
+    /// The ID of the box representing the entry-point of the query.
+    pub top_box: BoxId,
+    /// All boxes in the query graph model.
+    boxes: HashMap<BoxId, Box<RefCell<QueryBox>>>,
+    /// Used for asigning unique IDs to query boxes.
+    next_box_id: usize,
+    /// All quantifiers in the query graph model.
+    quantifiers: HashMap<QuantifierId, Box<RefCell<Quantifier>>>,
+    /// Used for asigning unique IDs to quantifiers.
+    next_quantifier_id: usize,
+}
+
+impl Model {
+    fn new() -> Self {
+        Self {
+            top_box: 0,
+            boxes: HashMap::new(),
+            next_box_id: 0,
+            quantifiers: HashMap::new(),
+            next_quantifier_id: 0,
+        }
+    }
+
+    fn make_box(&mut self, box_type: BoxType) -> BoxId {
+        let id = self.next_box_id;
+        self.next_box_id += 1;
+        let b = Box::new(RefCell::new(QueryBox {
+            id,
+            box_type,
+            columns: Vec::new(),
+            quantifiers: QuantifierSet::new(),
+            ranging_quantifiers: QuantifierSet::new(),
+            distinct: DistinctOperation::Preserve,
+        }));
+        self.boxes.insert(id, b);
+        id
+    }
+
+    fn make_select_box(&mut self) -> BoxId {
+        self.make_box(BoxType::Select(Select::new()))
+    }
+
+    fn get_box(&self, box_id: BoxId) -> &RefCell<QueryBox> {
+        &*self.boxes.get(&box_id).unwrap()
+    }
+
+    /// Create a new quantifier and adds it to the parent box
+    fn make_quantifier(
+        &mut self,
+        quantifier_type: QuantifierType,
+        input_box: BoxId,
+        parent_box: BoxId,
+    ) -> QuantifierId {
+        let id = self.next_quantifier_id;
+        self.next_quantifier_id += 1;
+        let q = Box::new(RefCell::new(Quantifier {
+            id,
+            quantifier_type,
+            input_box,
+            parent_box,
+            alias: None,
+        }));
+        self.quantifiers.insert(id, q);
+        self.get_box(parent_box).borrow_mut().quantifiers.insert(id);
+        self.get_box(input_box)
+            .borrow_mut()
+            .ranging_quantifiers
+            .insert(id);
+        id
+    }
+
+    fn get_quantifier(&self, quantifier_id: BoxId) -> &RefCell<Quantifier> {
```

Comment:
Why should we require a reference to the value instead of the value itself? We are interested in the value.

Reply:
This is similar to the signature of the `HashMap::get` method:

```rust
pub fn get<Q: ?Sized>(&self, k: &Q) -> Option<&V>
where
    K: Borrow<Q>,
    Q: Hash + Eq, 
```

At the moment all three usage of the `get_quantifier` method need explicit of the `quantifier_id` argument at the call site. It will be a bit simpler if you just accept `&BoxId` and do the dereferencing inside `get_quantifier`. This also allows you to pass references to types that implement `Borrow<BoxId>` directly as arguments.